File: monadic-composition.ts
import { MathFuncComposeReturnArray, MathFuncOneParamReturnArray } from "./types"

const add5:MathFuncOneParamReturnArray = (x) =>[ x + 5, '+ 5']
const double:MathFuncOneParamReturnArray = (x) =>[ x * 2, '* 2']
const deduct3:MathFuncOneParamReturnArray =(x) => [x - 3, '- 3']

const compose: MathFuncComposeReturnArray = (...funcs) => (x) => {
	return funcs.reduce(([x, desc], func) =>{
		const [x2, desc2] = func(x)
		return [x2, `${desc} ${desc2}`]
	}, [x, 'x'])
}

const app = compose(add5,deduct3, double)
console.log(app(10))

File: monadic-composition.dart
import 'types.dart';

MathFunctionOneParamReturnMap add5 = (x) => {x + 5: '+ 5'};
MathFunctionOneParamReturnMap double = (x) => {x * 2: '* 2'};
MathFunctionOneParamReturnMap subtract3 = (x) => {x - 3: '- 3'};

final MathFunctionOneParamComposeListMap compose = (funcs) {
  return (int x) => funcs.fold({x: 'x'}, (acc, fn) {
        final entry = acc.entries.first;
        final res = fn(entry.key);
        final newEntry = res.entries.first;
        return {newEntry.key: '${entry.value} ${newEntry.value}'};
      });
};

void main(List<String> args) {
  final app = compose([add5, double, subtract3]);
  print(app(5));
}
